ztoc/ztoc_builder.go

```diff
@@ -17,238 +17,110 @@
 package ztoc
 
 import (
-	"archive/tar"
-	"compress/gzip"
 	"fmt"
-	"io"
-	"os"
-
-	"github.com/awslabs/soci-snapshotter/compression"
-	"github.com/opencontainers/go-digest"
 )
 
-func BuildZtoc(gzipFile string, span int64, buildToolIdentifier string) (*Ztoc, error) {
-	if gzipFile == "" {
-		return nil, fmt.Errorf("need to provide gzip file")
-	}
-
-	index, err := compression.NewGzipZinfoFromFile(gzipFile, span)
-	if err != nil {
-		return nil, err
-	}
-	defer index.Close()
-
-	fm, uncompressedArchiveSize, err := getGzipFileMetadata(gzipFile, index)
-	if err != nil {
-		return nil, err
-	}
-
-	fs, err := getFileSize(gzipFile)
-	if err != nil {
-		return nil, err
-	}
-
-	digests, err := getPerSpanDigests(gzipFile, int64(fs), index)
-	if err != nil {
-		return nil, err
-	}
+// Builder holds a single `TocBuilder` that builds toc, and one `ZinfoBuilder`
+// *per* compression algorithm that builds zinfo. `TocBuilder` is shared by different
+// compression algorithms. Which `ZinfoBuilder` is used depends on the compression
+// algorithm used by the layer.
+type Builder struct {
+	tocBuilder    TocBuilder
+	zinfoBuilders map[string]ZinfoBuilder
 
-	checkpoints, err := index.Bytes()
-	if err != nil {
-		return nil, err
-	}
-
-	toc := TOC{
-		Metadata: fm,
-	}
+	buildToolIdentifier string
+}
 
-	compressionInfo := CompressionInfo{
-		MaxSpanID:   index.MaxSpanID(),
-		SpanDigests: digests,
-		Checkpoints: checkpoints,
+// NewBuilder creates a `Builder` used to build ztocs. By default it supports gzip,
+// user can register new compression algorithms by calling `RegisterCompressionAlgorithm`.
+func NewBuilder(buildToolIdentifier string) *Builder {
```

Comment:
FYI @dvnguyen-amzn this moves `buildToolIdentifier` from `builder.Build` to `ztoc.NewBuilder`.